src/components/svg-canvas-graph.tsx

```diff
@@ -156,21 +162,76 @@ const SvgCanvas = () => {
         if (e.shiftKey && selected.has(edge)) dispatch(removeSelected(edge));
         else dispatch(addSelected(edge));
+
+        if (mode.startsWith('station') || mode.startsWith('misc-node-virtual')) {
+            const x = e.clientX - document.getElementById('canvas')!.getBoundingClientRect().left;
+            const y = e.clientY - document.getElementById('canvas')!.getBoundingClientRect().top;
+            // Add station in the current line
+            const isStation = mode.startsWith('station');
+            const rand = nanoid(10);
+            const id = isStation ? (`stn_${rand}` as StnId) : (`misc_node_${rand}` as MiscNodeId);
+            const stnType = isStation ? (mode.slice(8) as StationType) : (mode.slice(10) as MiscNodeType);
+            const { x: svgX, y: svgY } = pointerPosToSVGCoord(x, y, svgViewBoxZoom, svgViewBoxMin);
+            // deep copy to prevent mutual reference
+            const attr = isStation
+                ? structuredClone(stations[stnType as StationType].defaultAttrs)
+                : structuredClone(miscNodes[stnType as MiscNodeType].defaultAttrs);
+            if ('color' in attr) attr.color = theme;
+            graph.current.addNode(id, {
+                visible: true,
+                zIndex: 0,
+                x: roundToNearestN(svgX, 5),
+                y: roundToNearestN(svgY, 5),
+                type: stnType,
+                [stnType]: attr,
+            });
+
+            const zIndex = graph.current.getEdgeAttribute(edge, 'zIndex');
+            const linePathType = graph.current.getEdgeAttribute(edge, 'type');
+            const lineStyleType = graph.current.getEdgeAttribute(edge, 'style');
+            const [source, target] = graph.current.extremities(edge);
+            const typeAttr = graph.current.getEdgeAttributes(edge)[linePathType];
+            const styleAttr = graph.current.getEdgeAttributes(edge)[lineStyleType];
+            graph.current.addDirectedEdgeWithKey(`line_${nanoid(10)}`, source, id, {
+                visible: true,
+                zIndex,
+                type: linePathType,
+                [linePathType]: structuredClone(typeAttr),
+                style: lineStyleType,
+                [lineStyleType]: structuredClone(styleAttr),
+                reconcileId: '',
+            });
+            graph.current.addDirectedEdgeWithKey(`line_${nanoid(10)}`, id, target, {
+                visible: true,
+                zIndex,
+                type: linePathType,
+                [linePathType]: structuredClone(typeAttr),
+                style: lineStyleType,
+                [lineStyleType]: structuredClone(styleAttr),
+                reconcileId: '',
+            });
+            graph.current.dropEdge(edge);
+            refreshAndSave();
+            if (isAllowProjectTelemetry) rmgRuntime.event(Events.ADD_STATION, { stnType });
+            dispatch(setSelected(new Set([id])));
+        }
+        dispatch(setMode('free'));
```

Comment:
free mode should be set only when previous mode is `station` or `misc-node-virtual`?